src/Compilers/CSharp/Portable/Binder/Binder_InterpolatedString.cs

```diff
@@ -134,5 +127,329 @@ private BoundExpression BindInterpolatedString(InterpolatedStringExpressionSynta
+
+                // Prior to C# 10, all types in an interpolated string expression needed to be convertible to `object`. After 10, some types
+                // (such as Span<T>) that are not convertible to `object` are permissible as interpolated string components, provided there
+                // is an applicable AppendFormatted method that accepts them. To preserve langversion, we therefore make sure all components
+                // are convertible to object if the current langversion is lower than the interpolation feature
+                TypeSymbol? objectType = null;
+                BindingDiagnosticBag? conversionDiagnostics = null;
+                var needToCheckConversionToObject = !Compilation.IsFeatureEnabled(MessageID.IDS_FeatureImprovedInterpolatedStrings) && diagnostics.AccumulatesDiagnostics;
+                if (needToCheckConversionToObject)
+                {
+                    objectType = GetSpecialType(SpecialType.System_Object, diagnostics, unconvertedInterpolatedString.Syntax);
+                    conversionDiagnostics = BindingDiagnosticBag.GetInstance(template: diagnostics);
+                }
+
+                Debug.Assert(appendCalls.Length == unconvertedInterpolatedString.Parts.Length);
+                var partsBuilder = ArrayBuilder<BoundExpression>.GetInstance(appendCalls.Length);
+                int baseStringLength = 0;
+                int numFormatHoles = 0;
+
+                for (int i = 0; i < appendCalls.Length; i++)
+                {
+                    var currentFormatCall = appendCalls[i];
+                    Debug.Assert(currentFormatCall is { HasErrors: true } or BoundCall { Arguments: { Length: > 0 } } or BoundDynamicInvocation);
+                    var currentPart = unconvertedInterpolatedString.Parts[i];
+                    BoundExpression newPart;
+
+                    // If we were able to actually bind the Append call, then just update the current part to be the call.
+                    // Otherwise, there was some kind of error, and we just want to keep the original part. We'll update
+                    // the HasErrors to be true, but we don't want IOperation to have to try and find the original node
+                    // to include in the tree in a potentially arbitrary tree
+                    if (currentFormatCall is BoundCall call)
+                    {
+                        Debug.Assert(call.Arguments.Length > 0);
+                        newPart = currentPart is BoundStringInsert si
+                            ? si.Update(call, si.Alignment, si.Format, isInterpolatedStringBuilderAppendCall: true)
+                            : call;
+                    }
+                    else
+                    {
+                        Debug.Assert(currentFormatCall.HasErrors);
+                        newPart = currentPart.WithHasErrors();
```

Comment:
>currentPart [](http://example.com/codeflow?start=34&length=11)

If we are dropping the result of binding a call, It feels like at least we need to snap this node to natural type for error recovery.